[wasm][debugger] Small improvements to fail gracefully

radical opened the pull request:
- Try to fail gracefully when we come across some variable/value that we can't describe (yet!)
- Eg. missing because of invalid param/local id, or value description
failed because of some as yet unsupported type

Fixes https://github.com/dotnet/runtime/issues/41744
Fixes https://github.com/dotnet/runtime/issues/41818

radical committed:
[wasm][debugger] Instead of failing completely, skip the problematic

.. property. Some times we might not get a `value`, or `name`, or it
might instead have a `get`. Handle those cases correctly when combining
the name/value/get objects.

This showed up in case of a `MulticastDelegate`, where we didn't have a
`value`, and ended up incorrectly combining the name/value objects, thus
returning incorrect locals.

radical committed:
[wasm][debugger] Handle MulticastDelegate, and events

- Essentially we just surface these as a symbol showing the type name

radical committed:
[wasm][debugger] Fail gracefully when vt could not be expanded

radical commented:
This is based on #41562 .

radical commented:
You can see just the changes for this PR: https://github.com/radical/runtime/compare/wasm-boxing...radical:wasm-misc-fixes

thaystg commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -362,43 +362,97 @@ var MonoSupportLib = {
+			props.forEach(p => {
+				if (p.name === undefined) {
+					console.debug(`Bug: Found a member with no name. Skipping it. p: ${JSON.stringify(p)}`);
+					return;
```
Should we return or we can just continue the forEach for the next member?

radical replied:
It's a lambda, so return here will effectively mean - continue.

radical commented:
Opened an issue corresponding to this - https://github.com/dotnet/runtime/issues/41744

radical committed:
[wasm][debugger] Handle invalid scope ids

scope<0, or scope too high

- This does get filtered at the proxy level, but app side should be able
to handle it too

radical committed:
[wasm][debugger] Handle invalid/missing/failed value descriptions

- Eg. missing because of invalid param/local id, or value description
failed because of some as yet unsupported type

radical committed:
[wasm][debugger] Fix frame indexing

- `collect_frames`, `list_frames` - both iterate over stack frames. They
skip some frames. This debug proxy assigns a simple index to each of the
received(filtered) frames.

    - so, if we had `[ frame0, (skipped frame1), frame2 ]`, the proxy will
    have `[ frame0(index:0), frame2(index:1) ]`

- `describe_variables_on_frame` also iterates over stack frames, and
tries to find a given frame by an index. And this index is what the
proxy had assigned.
    - because some frames were skipped, this function also needs to skip
    the *same* ones, so that it can find the intended frame.

    - Instead of trying to keep these in sync, here the indexing is
    changed to be the real index found as we iterate over the frames.
    - And the proxy just uses that.
    - So, we will have `[ frame0(index:0), frame2(index:2) ]`

This showed up in a trace in aspnetcore, which was called via
reflection. And that frame didn't get added to the list because it was
not `MONO_WRAPPER_NONE`, which caused the indexing to get out of sync.

Fixes: https://github.com/dotnet/runtime/issues/41818

radical committed:
fix warning: remove unused var

radical commented:
Rebased again on https://github.com/dotnet/runtime/pull/41562 . 
Changes specific to this PR - https://github.com/radical/runtime/compare/wasm-boxing...radical:wasm-misc-fixes

radical committed:
rebase on master, fix errors

radical commented:
This is rebased on master now, so can be reviewed independently.

lewing reviewed (commented):
the frame_id logic is getting a little odd now, it appears to start at 1 in the mono response but gets reduced to 0 in the debugger response?  have you verified that stack inspection works properly when selecting different frames (we do test for this now right?)

radical commented:
> the frame_id logic is getting a little odd now, it appears to start at 1 in the mono response but gets reduced to 0 in the debugger response? have you verified that stack inspection works properly when selecting different frames (we do test for this now right?)

MonoProxy is the one doing `-1`, the IDE side will always see that final number only. But I agree it is a little.. odd. I can fix it up in debugger.c itself.
Yeah, I did test for this. And we do have some tests for it.

radical commented:
> > the frame_id logic is getting a little odd now, it appears to start at 1 in the mono response but gets reduced to 0 in the debugger response? have you verified that stack inspection works properly when selecting different frames (we do test for this now right?)
> 
> MonoProxy is the one doing `-1`, the IDE side will always see that final number only. But I agree it is a little.. odd. I can fix it up in debugger.c itself.
> Yeah, I did test for this. And we do have some tests for it.

And that commit adds a test specifically for checking a previous frame. In the case of the test, we skip some frames too, so discrepancies in indexing should show up.

radical committed:
Make frame indices returned from debugger.c, 0-based

- Earlier this 1-based, and it was being adjusted in `MonoProxy`.
- Based on @lewing's suggestion, changing this to be 0-based in
debugger.c, itself, thus removing the need to "fixup" in `MonoProxy`.

radical committed:
dotnet-format fixes

radical committed:
Merge remote-tracking branch 'origin/master' into wasm-misc-fixes
